Add Car Cleaning Service functionality for Assignment 2

cristichirtos opened the pull request:
This pull request contains the whole functionality of the Car Cleaning Service for the second assignment.

The changes in this PR are:
- Implemented MVC pattern
- Added Car and CarCleaner models in the model folder
- Added a console view in the view folder
- Added a controller for the console application in the controller folder
- Added a launcher script in the root directory
- Implemented a layered architecture with models services and repositories (although since there is no database, there are no repositories)
- Added CarCleanerService service in the service folder which handles all the functionality of the car cleaner. 
- Added all tests for the models, the service and the validator.

vaneaclaudia commented on car_cleaning_service_with_threads.rb:
```diff
@@ -0,0 +1,115 @@
+  end
+
+  def to_s
+    @id
```
If you use `attr_reader` then you don't need to use `@id` - you can replace it with `id`. Also I would not recommend overwriting  `to_s` and `to_sym` methods in general because those are methods from ruby `Object` class

cristichirtos replied:
Oh, I didn't realize that the @ is no longer required, but it makes sense! I overwrote the to_sym method to be able to use symbols as keys for the hash map containing Car objects, instead of the object itself. Is that a bad practice?

vaneaclaudia replied:
Regarding overriding, it is not bad but I recommend to use it with caution in general. For example if you are overriding `to_s` method globally in a monkey patching then it might break other parts of code which relates to this method. So the comment was more like a note to be aware of this.

In your specific case `to_s` method returns just the id, without modifying it. So you could just use car.id instead of car.to_s

cristichirtos replied:
I changed the implementation, thank you for the suggestion! It looks cleaner now.

vaneaclaudia commented on car_cleaning_service_with_threads.rb:
```diff
@@ -0,0 +1,115 @@
+  public
+
+  def add_car(car)
+    @queue.push(car)
```
you can add `queue` to `attr_reader` and use it without @ symbol

vaneaclaudia commented on car_cleaning_service_with_threads.rb:
```diff
@@ -0,0 +1,115 @@
+    self.init_threads
+  end
+
+  public
```
why do you need to specify `public` here?

cristichirtos replied:
I only specified it for easier finding of the public methods, but I have no need for it as it is implied here.

vaneaclaudia commented on car_cleaning_service_with_threads.rb:
```diff
@@ -0,0 +1,115 @@
+  end
+
+  def pick_up_car(car)
+    if @ready.has_key?(car.to_sym)
```
you can add `ready` to `attr_reader` and use it without @ symbol. It will increase readability

cristichirtos replied:
Thank you!

vaneaclaudia commented on car_cleaning_service_without_threads.rb:
```diff
@@ -0,0 +1,198 @@
+require 'time'
+
+class Car
```
you can extract `Car` class intro a separate class since you are reusing it

cristichirtos replied:
I am refactoring the code to have multiple files, as it is easier to work with cleaner code, so I will extract it and move the classes into packages.

vaneaclaudia commented on car_cleaning_service_without_threads.rb:
```diff
@@ -0,0 +1,198 @@
+  end
+end
+
+class CarValidator
```
nice idea to have a `CarValidator` 👍

cristichirtos replied:
Thanks!

vaneaclaudia commented on car_cleaning_service_without_threads.rb:
```diff
@@ -0,0 +1,198 @@
+  def valid?
+    validate_id
+    validate_time
+    @errors.empty?
```
you have `attr_reader ` for errors object. You can use it without @ symbol

vaneaclaudia commented on car_cleaning_service_without_threads.rb:
```diff
@@ -0,0 +1,198 @@
+
+  public
+
+  def add_car(car)
```
please add some empty lines between code blocks to increase readability

cristichirtos replied:
Will do!

vaneaclaudia commented on car_cleaning_service_without_threads.rb:
```diff
@@ -0,0 +1,198 @@
+    # if the client did not specify the time they need the car, I consider it to be the end of the day
+    car.required_time = Time.parse(@time.strftime("%Y-%m-%dT#{today_closing_hour}:00:00%z")) if car.required_time.nil?
+    position_in_queue = -1
+    @queue.each { |car_in_queue|
```
use `do...end` for multi lines blocks and `{}` when there is a single line block

vaneaclaudia commented on car_cleaning_service_without_threads.rb:
```diff
@@ -0,0 +1,198 @@
+
+  def handle_car_servicing
+    anything_happened = false
+    @cars_in_process.each { |car, enter_time|
```
same comment as above for multi line blocks

vaneaclaudia commented on car_cleaning_service_without_threads.rb:
```diff
@@ -0,0 +1,198 @@
+  def today_closing_hour
+    return SATURDAY_CLOSING_HOUR if @time.saturday?
+    return SUNDAY_CLOSING_HOUR if @time.sunday?
+    return WEEKDAY_CLOSING_HOUR
```
no need for `return` here since is the last line of the method.

vaneaclaudia commented on car_cleaning_service_with_threads.rb:
```diff
@@ -0,0 +1,115 @@
+
+  attr_reader :time
+
+  def initialize(queue = Queue.new)
```
what if you pass the Car object here in the initializer ? In this way you will get rid of the car parameter passed to the methods.

cristichirtos replied:
I will reformat the code such that the CarCleaningService will be split into a model (the CarCleaner itself, holding only the data) and a service (CarCleanerService, implementing the functionality related to the CarCleaner) and I cannot pass the Car object itself to the CarCleaner as it holds a list of cars that enter this facility, and initially it is empty. As cars arrive, they are inserted into the queue based on the time they are needed (the queue is an array which behaves like a priority queue).

vaneaclaudia replied:
Awesome, let's see how it will look like afterwards.

vaneaclaudia commented on car_cleaning_service_with_threads.rb:
```diff
@@ -0,0 +1,115 @@
+  private
+
+  def init_threads
+    @threads = MAX_CARS_AT_ONCE.times.map {
```
use `do...end` for multi lines blocks and `{} `when there is a single line block

cristichirtos replied:
Thank you!

vaneaclaudia commented on car_cleaning_service_without_threads.rb:
```diff
@@ -0,0 +1,198 @@
+  end
+
+  def simulate_service(number_of_hours)
+    number_of_hours.times {
```
same comment as above for multi line blocks

cristichirtos commented:
@vaneaclaudia @VictorMotogna I have refactored the code (pretty much everything is changed) and I implemented an MVC pattern as I find it easier to test, control and extend the functionality with it. Also separated the code in layers. Let me know what you think! (Test are coming soon)

vaneaclaudia commented on controller/car_cleaning_system_controller.rb:
```diff
@@ -0,0 +1,89 @@
+    car_cleaning_system_view.print_result(result)
+  end
+
+  def on_time()
```
no need for empty () if there are no parameters to the method.

cristichirtos replied:
I wrote all those methods at once and I forgot to remove the paranthesis on the ones that require no params. Thanks!

vaneaclaudia commented on controller/car_cleaning_system_controller.rb:
```diff
@@ -0,0 +1,89 @@
+    car_cleaning_system_view.print_result(result)
+  end
+
+  def on_queue()
```
same comment as above

vaneaclaudia commented on service/car_cleaner_service.rb:
```diff
@@ -0,0 +1,120 @@
+      true
+    else
+      false
+    end
```
another way of writing these lines of code with 3 lines:

```
return false unless car_cleaner.cars_ready.has_key?(car_id.to_sym)

car_cleaner.cars_ready.delete(car_id.to_sym)
true
```

cristichirtos replied:
Nice, thanks!

vaneaclaudia commented on service/car_cleaner_service.rb:
```diff
@@ -0,0 +1,120 @@
+    time_until_processing = (car_cleaner.queue.index(car) / CarCleaner::MAX_CARS_AT_ONCE) * CarCleaner::CAR_CLEANING_TIME_IN_HOURS
+
+    until time_until_processing == 0
+
```
Regarding this empty line at the beginning of the block, there is no need for this since you are just starting the block so you don't have a pice of code to delimitate from another. Same remark for all the empty lines you have at the beginning of the blocks.

cristichirtos replied:
I understand, I will remove them.

vaneaclaudia commented on service/car_cleaner_service.rb:
```diff
@@ -0,0 +1,120 @@
+      estimated_time += ONE_HOUR_IN_SECONDS
+    end
+
+    return format_time(estimated_time)
```
redundant return since it's the last line of the method, the result for `format_time(estimated_time)` will be anyway returned

vaneaclaudia commented on service/car_cleaner_service.rb:
```diff
@@ -0,0 +1,120 @@
+      anything_happened = true
+    end
+
+    return anything_happened
```
same remark regarding redundant return

cristichirtos replied:
I wanted to ask if we should include the return keyword to explicitly show what the methods return to make it easier to understand, but I figured from your comments that it's not necessary. Thanks!

vaneaclaudia commented on spec/model/car_cleaner_spec.rb:
```diff
@@ -0,0 +1,13 @@
+  let(:car_cleaner_empty) { CarCleaner.new }
+  let(:car_cleaner_with_data) { CarCleaner.new(["Test", "Another Test"]) }
+  
+  it "stores an array as a queue" do
```
use single quotes where you don't have interpolation or where you don't need to use double quotes

`it 'stores an array as a queue'`

vaneaclaudia commented on spec/model/car_cleaner_spec.rb:
```diff
@@ -0,0 +1,13 @@
+    expect(car_cleaner_empty.queue).to eq []
+    expect(car_cleaner_with_data.queue). to eq ["Test", "Another Test"]
+  end
+
```
I'll write you here how I would have done these tests:

```
context 'with empty data' do
  subject { CarCleaner.new }

  it 'stores an array as a queue' do
    response = subject

    expect(response.queue).to eq []
  end
end


context 'with data' do
  subject { CarCleaner.new(['Test', 'Another Test']) }

  it 'stores an array as a queue' do
    response = subject

    expect(response.queue). to eq ['Test', 'Another Test']
  end
end
```

Since you are testing 2 different behaviours it is ok to have 2 different contexts instead of testing 2 behaviours in the same it. Let me know if it's not clear enough.

cristichirtos replied:
It's clear and it makes more sense. Thank you!

vaneaclaudia commented on spec/model/car_spec.rb:
```diff
@@ -0,0 +1,11 @@
+describe Car do
+  let(:car) { Car.new("CJ30PXP", "2020-11-24") }
+  
+  it "stores an id and a date" do
```
same remark here regarding double quotes. You can write all the strings from this class by using single quotes

vaneaclaudia commented on spec/service/car_cleaner_service_spec.rb:
```diff
@@ -0,0 +1,53 @@
+require 'timecop'
+
+describe CarCleanerService do
+  before { Timecop.freeze(Time.local(2021, 4, 20, 8, 0, 0)) }
```
nice 👍

cristichirtos replied:
Thank you!

vaneaclaudia commented on spec/service/car_cleaner_service_spec.rb:
```diff
@@ -0,0 +1,53 @@
+describe CarCleanerService do
+  before { Timecop.freeze(Time.local(2021, 4, 20, 8, 0, 0)) }
+  let(:car_cleaner_service) { CarCleanerService.new Time.now }
+  
```
my suggestion is to try to be consistent in the way you are calling the methods, for example here you are using `CarCleanerService.new Time.now ` without parentheses but then you also use `Car.new("CJ29TST", "2021-04-21")` with parentheses. This is a general comment, I noticed through the code that both ways are used.

cristichirtos replied:
I sometimes don't use paranthesis if there is only one argument, but indeed it can get confusing. I'll use paranthesis everywhere.

vaneaclaudia commented on spec/service/car_cleaner_service_spec.rb:
```diff
@@ -0,0 +1,53 @@
+    let(:car) { Car.new "CJ30TST"}
+
+    it "adds a car in the queue" do
+
```
same comment here regarding blank line at the beginning of the block

vaneaclaudia reviewed (approved):
Good job 👍 👏

cristichirtos commented:
Thank you!

VictorMotogna commented on service/car_cleaner_service.rb:
```diff
@@ -0,0 +1,114 @@
+
+  def add_car(car)
+    # if the client did not specify the time they need the car, I consider it to be the end of the day
+    car.required_time = Time.parse(time.strftime("%Y-%m-%dT#{today_closing_hour}:00:00%z")) if car.required_time.nil?
```
you could remove this comment and do something like this, to be more intuitive and easier to understand:
```ruby
end_of_day = Time.parse(time.strftime("%Y-%m-%dT#{today_closing_hour}:00:00%z"))
car.required_time = end_of_day if car.required_time.nil?
```

cristichirtos replied:
That's a good idea, thanks!

VictorMotogna commented on service/car_cleaner_service.rb:
```diff
@@ -0,0 +1,114 @@
+
+  def pick_up_car(car_id)
+    return false unless car_cleaner.cars_ready.has_key?(car_id.to_sym)
+    car_cleaner.cars_ready.delete(car_id.to_sym)
```
if we are having a breaking line (like `return`, in this case), it's a good practice to add a newline between it and the next block of the code

cristichirtos replied:
Makes sense, I will add them where applicable.

VictorMotogna commented on service/car_cleaner_service.rb:
```diff
@@ -0,0 +1,114 @@
+    format_time(estimated_time)
+  end
+
+  def format_time(time_asked = @time)
```
I think that if you have time as `attr_accessor` or `attr_reader` you do not need the `@`. For example:
```ruby
class TestClass
  attr_accessor :time

  def initialize(time)
    @time = time
  end

  def print_time(this_time = time)
    this_time
  end
end

puts TestClass.new(Time.now).print_time # prints 2021-04-21 16:27:51 +0300
puts TestClass.new(Time.now).print_time(Time.new(2018, 1, 1)) # prints 2018-01-01 00:00:00 +0200
```

cristichirtos replied:
You are right, I modified it.

VictorMotogna commented on service/car_cleaner_service.rb:
```diff
@@ -0,0 +1,114 @@
+  def today_closing_hour
+    return CarCleaner::SATURDAY_CLOSING_HOUR if @time.saturday?
+    return CarCleaner::SUNDAY_CLOSING_HOUR if @time.sunday?
+    CarCleaner::WEEKDAY_CLOSING_HOUR
```
same here, try to add a newline after `returns`

VictorMotogna commented on spec/model/car_cleaner_spec.rb:
```diff
@@ -0,0 +1,27 @@
+
+  end
+  
+end
```
I noticed you use to add a newline before the last `end`. I don't think that's necessary. it's not wrong or anything, but it's not necessary and rubocop will complain about that

cristichirtos replied:
I will remove the redundant lines.

VictorMotogna commented on spec/model/validator/car_validator_spec.rb:
```diff
@@ -0,0 +1,59 @@
+require_relative '../../../model/validator/car_validator'
+
+describe CarValidator do
+  before { Timecop.freeze(Time.local(2021, 4, 20, 8, 0, 0)) }
```
nice 👍

cristichirtos replied:
Thanks!

cristichirtos commented:
@VictorMotogna I implemented the changes everywhere in the code were they were applicable. Let me know what you think!
